src/OpenApi/src/Transformers/Implementations/OpenApiSchemaReferenceTransformer.cs

```diff
@@ -0,0 +1,137 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.AspNetCore.OpenApi;
+
+/// <summary>
+/// Document transformer to support mapping duplicate JSON schema instances
+/// into JSON schema references across the document.
+/// </summary>
+internal sealed class OpenApiSchemaReferenceTransformer : IOpenApiDocumentTransformer
+{
+    public Task TransformAsync(OpenApiDocument document, OpenApiDocumentTransformerContext context, CancellationToken cancellationToken)
+    {
+        var schemaStore = context.ApplicationServices.GetRequiredKeyedService<OpenApiSchemaStore>(context.DocumentName);
+        var schemasByReference = schemaStore.SchemasByReference;
+
+        foreach (var pathItem in document.Paths.Values)
+        {
+            for (var i = 0; i < OpenApiConstants.OperationTypes.Length; i++)
```

Comment:
Was it not possible to `foreach` over `OpenApiConstants.OperationTypes`?

Reply:
It is -- I copied this over from the implementation in `DelegateOpenTransformer` which is aggressive about not allocating enumerators 🤷🏽‍♀️